File: Windows/src/CodexMeter.Core/Domain/TokenUsage.cs
namespace CodexMeter.Core.Domain;

public readonly record struct TokenUsage(
    long InputTokens,
    long CachedInputTokens,
    long OutputTokens)
{
    public static TokenUsage Zero { get; } = new(0, 0, 0);

    public long TotalTokens => SaturatingAdd(
        SaturatingAdd(InputTokens, CachedInputTokens),
        OutputTokens);

    public bool IsValid => InputTokens >= 0
        && CachedInputTokens >= 0
        && OutputTokens >= 0
        && CachedInputTokens <= InputTokens;

    public TokenUsage Add(TokenUsage other) => new(
        SaturatingAdd(InputTokens, other.InputTokens),
        SaturatingAdd(CachedInputTokens, other.CachedInputTokens),
        SaturatingAdd(OutputTokens, other.OutputTokens));

    public TokenUsage SubtractFloorAtZero(TokenUsage previous) => new(
        Math.Max(0, InputTokens - previous.InputTokens),
        Math.Max(0, CachedInputTokens - previous.CachedInputTokens),
        Math.Max(0, OutputTokens - previous.OutputTokens));

    public TokenUsage ComponentWiseMaximum(TokenUsage other) => new(
        Math.Max(InputTokens, other.InputTokens),
        Math.Max(CachedInputTokens, other.CachedInputTokens),
        Math.Max(OutputTokens, other.OutputTokens));

    public bool IsComponentWiseAtLeast(TokenUsage other) =>
        InputTokens >= other.InputTokens
        && CachedInputTokens >= other.CachedInputTokens
        && OutputTokens >= other.OutputTokens;

    private static long SaturatingAdd(long left, long right)
    {
        try
        {
            return checked(left + right);
        }
        catch (OverflowException)
        {
            return long.MaxValue;
        }
    }
}

File: Sources/CodexMeter/Domain/TokenUsage.swift
import Foundation

struct TokenUsage: Codable, Equatable, Sendable {
    let inputTokens: Int64
    let cachedInputTokens: Int64
    let outputTokens: Int64

    static let zero = TokenUsage(inputTokens: 0, cachedInputTokens: 0, outputTokens: 0)

    var totalTokens: Int64 {
        inputTokens
            .saturatedAdding(cachedInputTokens)
            .saturatedAdding(outputTokens)
    }

    var isValid: Bool {
        inputTokens >= 0
            && cachedInputTokens >= 0
            && outputTokens >= 0
            && cachedInputTokens <= inputTokens
    }

    func adding(_ other: TokenUsage) -> TokenUsage {
        TokenUsage(
            inputTokens: inputTokens.saturatedAdding(other.inputTokens),
            cachedInputTokens: cachedInputTokens.saturatedAdding(other.cachedInputTokens),
            outputTokens: outputTokens.saturatedAdding(other.outputTokens)
        )
    }

    func subtractingFloorAtZero(_ previous: TokenUsage) -> TokenUsage {
        TokenUsage(
            inputTokens: max(0, inputTokens - previous.inputTokens),
            cachedInputTokens: max(0, cachedInputTokens - previous.cachedInputTokens),
            outputTokens: max(0, outputTokens - previous.outputTokens)
        )
    }

    func componentWiseMaximum(with other: TokenUsage) -> TokenUsage {
        TokenUsage(
            inputTokens: max(inputTokens, other.inputTokens),
            cachedInputTokens: max(cachedInputTokens, other.cachedInputTokens),
            outputTokens: max(outputTokens, other.outputTokens)
        )
    }

    func isComponentWiseAtLeast(_ other: TokenUsage) -> Bool {
        inputTokens >= other.inputTokens
            && cachedInputTokens >= other.cachedInputTokens
            && outputTokens >= other.outputTokens
    }
}

private extension Int64 {
    func saturatedAdding(_ other: Int64) -> Int64 {
        let result = addingReportingOverflow(other)
        return result.overflow ? Int64.max : result.partialValue
    }
}
